browser/brave_wallet/cardano_provider_renderer_browsertest.cc

```diff
@@ -0,0 +1,1364 @@
+#include "third_party/abseil-cpp/absl/strings/str_format.h"
+#include "ui/base/l10n/l10n_util.h"
+
+namespace brave_wallet {
+
+namespace {
+
+using testing::_;
+
+constexpr char kCheckCardanoProviderScript[] =
+    "!!window.cardano && !!window.cardano.brave";
+constexpr char kOverwriteCardanoScript[] =
+    "window.cardano = ['test']; window.cardano[0]";
+constexpr char kOverwriteCardanoBraveScript[] =
+    "window.cardano.brave = ['test']; window.cardano.brave[0]";
+constexpr char kExtensionWallet[] = "window.cardano.somewallet = ['test'];";
+constexpr char kCheckExtensionWallet[] =
+    "!!window.cardano && !!window.cardano.somewallet";
+
+std::string NonWriteableScriptMethod(std::string_view provider,
+                                     std::string_view method) {
+  return absl::StrFormat(
+      R"(new Promise(resolve => {
+          window.%s.%s = "brave"
+          if (typeof window.%s.%s === "function")
+            resolve(true);
+          else
+            resolve(false);
+          });
+        )",
+      provider, method, provider, method);
+}
+
+std::string NonWriteableScriptApiMethod(std::string_view method) {
+  return absl::StrFormat(
+      R"(async function check() {
+          let x = await window.cardano.brave.enable()
+          x.%s = "brave"
+          if (typeof x.%s === "function")
+            return true;
+          else
+            return false;
+          }
+          check();
+        )",
+      method, method);
+}
+
+std::string EnableScript() {
+  return
+      R"(async function connect() {
+          try {
+            const result = await window.cardano.brave.enable();
+            return true;
+          } catch (err) {
+            return false;
+          }
+        }
+        connect();)";
+}
+
+std::string NonWriteableScriptProperty(const std::string& property) {
+  return base::StringPrintf(
+      R"(window.cardano.brave.%s = "brave";
+         !(window.cardano.brave.%s === "brave");)",
+      property.c_str(), property.c_str());
+}
+
+class TestCardanoProvider : public brave_wallet::mojom::CardanoProvider {
+ public:
+  TestCardanoProvider() = default;
+  ~TestCardanoProvider() override = default;
+
+  MOCK_METHOD1(Enable, void(EnableCallback));
+  MOCK_METHOD1(IsEnabled, void(IsEnabledCallback));
+
+  MOCK_METHOD1(GetNetworkId, void(GetNetworkIdCallback));
+  MOCK_METHOD1(GetUsedAddresses, void(GetUsedAddressesCallback callback));
+  MOCK_METHOD1(GetUnusedAddresses, void(GetUnusedAddressesCallback callback));
+  MOCK_METHOD1(GetChangeAddress, void(GetChangeAddressCallback callback));
+  MOCK_METHOD1(GetRewardAddresses, void(GetRewardAddressesCallback callback));
+  MOCK_METHOD1(GetBalance, void(GetBalanceCallback callback));
+  MOCK_METHOD3(GetUtxos,
+               void(const std::optional<std::string>& amount,
+                    mojom::CardanoProviderPaginationPtr paginate,
+                    GetUtxosCallback callback));
+  MOCK_METHOD3(SignTx,
```

Comment:
We should not test mocked implementation if it is a browser test. Let's have a comment to remove that mock when core part is ready

Reply:
Mocks are fine when used for easier setup of some secondary-level stuff.
Browser tests should test browser as close as possible to production runtime, especially  the code which is being tested.

Reply:
https://github.com/brave/brave-browser/issues/46742